File: frontend/src/types/zalo.ts
export type ZaloChannel = 'oa_message' | 'zns'
export type ZaloLogStatus = 'pending' | 'sent' | 'failed' | 'delivered'

export const NOTIF_TYPE_LABELS: Record<string, string> = {
  task_overdue:  '⚠️ Nhiệm vụ quá hạn',
  task_warning:  '⏰ Nhiệm vụ sắp đến hạn',
  kpi_low:       '📊 KPI thấp',
  report_done:   '📋 Báo cáo hoàn thành',
  directive_new: '📌 Chỉ đạo mới',
  document_new:  '📄 Văn bản mới',
  system_alert:  '🔔 Cảnh báo hệ thống',
  broadcast:     '📢 Thông báo điều hành',
}

export const STATUS_LABELS: Record<ZaloLogStatus, string> = {
  pending:   'Chờ gửi',
  sent:      'Đã gửi',
  failed:    'Thất bại',
  delivered: 'Đã nhận',
}

export interface ZaloConfigRead {
  id: number
  app_id: string
  oa_id: string
  has_access_token: boolean
  has_refresh_token: boolean
  token_expiry: string | null
  is_active: boolean
  created_at: string
}

export interface ZaloConfigUpsert {
  app_id: string
  app_secret: string
  oa_id: string
  access_token?: string
  refresh_token?: string
  is_active?: boolean
}

export interface ZaloTemplateRead {
  id: number
  name: string
  notif_type: string
  channel: ZaloChannel
  subject: string
  content: string
  variables: string[] | null
  zns_template_id: string | null
  is_active: boolean
  is_default: boolean
  created_at: string
}

export interface ZaloTemplateCreate {
  name: string
  notif_type: string
  channel?: ZaloChannel
  subject: string
  content: string
  variables?: string[]
  zns_template_id?: string
  is_active?: boolean
}

export interface ZaloTemplateUpdate {
  name?: string
  channel?: ZaloChannel
  subject?: string
  content?: string
  variables?: string[]
  zns_template_id?: string
  is_active?: boolean
}

export interface ZaloLogRead {
  id: number
  template_id: number | null
  recipient_user_id: number | null
  recipient_phone: string
  notif_type: string
  subject: string
  content_rendered: string
  status: ZaloLogStatus
  error_msg: string | null
  zalo_msg_id: string | null
  triggered_by: string
  entity_type: string | null
  entity_id: number | null
  sent_at: string | null
  created_at: string
}

export interface ZaloUserLinkRead {
  id: number
  user_id: number
  zalo_phone: string | null
  zalo_user_id: string | null
  is_verified: boolean
  is_active: boolean
  created_at: string
}

export interface ZaloUserLinkUpsert {
  user_id: number
  zalo_phone?: string
  zalo_user_id?: string
  is_active?: boolean
}

export interface ZaloSendRequest {
  notif_type: string
  recipient_user_ids: number[]
  context?: Record<string, string>
  entity_type?: string
  entity_id?: number
}

export interface ZaloStats {
  total_sent: number
  sent_today: number
  failed_today: number
  users_linked: number
}

File: backend/app/schemas/zalo.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel


# ── ZaloConfig ────────────────────────────────────────────────────────────────

class ZaloConfigUpsert(BaseModel):
    app_id: str
    app_secret: str
    oa_id: str
    access_token: str = ""
    refresh_token: str = ""
    is_active: bool = True


class ZaloConfigRead(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    app_id: str
    oa_id: str
    has_access_token: bool = False
    has_refresh_token: bool = False
    token_expiry: datetime | None
    is_active: bool
    created_at: datetime


# ── ZaloTemplate ──────────────────────────────────────────────────────────────

class ZaloTemplateCreate(BaseModel):
    name: str
    notif_type: str
    channel: str = "oa_message"
    subject: str
    content: str
    variables: list[str] | None = None
    zns_template_id: str | None = None
    is_active: bool = True


class ZaloTemplateUpdate(BaseModel):
    name: str | None = None
    channel: str | None = None
    subject: str | None = None
    content: str | None = None
    variables: list[str] | None = None
    zns_template_id: str | None = None
    is_active: bool | None = None


class ZaloTemplateRead(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    name: str
    notif_type: str
    channel: str
    subject: str
    content: str
    variables: list | None
    zns_template_id: str | None
    is_active: bool
    is_default: bool
    created_at: datetime


# ── ZaloLog ───────────────────────────────────────────────────────────────────

class ZaloLogRead(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    template_id: int | None
    recipient_user_id: int | None
    recipient_phone: str
    notif_type: str
    subject: str
    content_rendered: str
    status: str
    error_msg: str | None
    zalo_msg_id: str | None
    triggered_by: str
    entity_type: str | None
    entity_id: int | None
    sent_at: datetime | None
    created_at: datetime


# ── ZaloUserLink ──────────────────────────────────────────────────────────────

class ZaloUserLinkUpsert(BaseModel):
    user_id: int
    zalo_phone: str | None = None
    zalo_user_id: str | None = None
    is_active: bool = True


class ZaloUserLinkRead(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    user_id: int
    zalo_phone: str | None
    zalo_user_id: str | None
    is_verified: bool
    is_active: bool
    created_at: datetime


# ── Manual send ───────────────────────────────────────────────────────────────

class ZaloSendRequest(BaseModel):
    notif_type: str
    recipient_user_ids: list[int]
    context: dict[str, Any] = {}
    entity_type: str | None = None
    entity_id: int | None = None


# ── Stats ─────────────────────────────────────────────────────────────────────

class ZaloStats(BaseModel):
    total_sent: int
    sent_today: int
    failed_today: int
    users_linked: int
    pending_conflicts: int = 0
